### ISSUE TITLE
Make new training sets for Lode Runner

### ISSUE THREAD

**Issue by schrum2** [schrum2/MM-NEAT#525]
Given what the GAN models are producing, I'm a little sad we're not seeing results that look as cool as those in the original game. After all, look at this diversity of levels:

![All Levels](https://user-images.githubusercontent.com/8397236/84691088-2b0c1500-af09-11ea-8c96-f49de1656569.png)

However, we know that if we train on a small set, that we can easily reproduce those levels and similar ones in between. Therefore, I want to train lots of models on small data sets (so small that you can even train easily without a CUDA device).

One approach is to simply train on groups of 10 levels: 1-10, 11-20, 21-30, etc. Do this and name each model appropriately. Start this training right away so it can happen in the background.

However, the other, more interesting approach, will require analyzing the levels a bit more. Specifically, we can train models exclusively on levels that have some kind of similar features. However, this will require categorizing the existing levels a bit more. 

Specifically, write some code that provides a profile on each level analyzing the following features:
- Percentage of empty tiles
- Percentage of ladder tiles
- Percentage of ground tiles (solid or diggable)
- Percentage of solid tiles
- Percentage of diggable tiles
- Connectivity score
- Length of A* path
- Percent backtracking in A* path (this is the percentage of locations on the A* path that are visited more than once)

Some of these are already calculated as fitness functions. You should extract that code into a new LodeRunnerLevelAnalysisUtil class so that you can call it from both the LodeRunnerLevelTask and from the new class you write for performing this analysis (or it could be a main method in LodeRunnerLevelAnalysisUtil). Also, note that many of these are calculating percentages ... you should be able to write one general method that accepts an array of the types of tiles to look for, and performs the calculation based on that. Then the more specific methods (like `percentEmptyTiles`) can be written with just one line of code that calls the more general method for calculating percentages.

Once the analysis is done, we will figure out what others sorts of models we want to train.

**Comment by schrum2** [comment]
By the way, let's stick to a latent vector length of 10 on all of these models

**Comment by kdsteckel** [comment]
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel1to10.json --experiment LodeRunnerAllTilesLevel1to10-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel11to20.json --experiment LodeRunnerAllTilesLevel11to20-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel21to30.json --experiment LodeRunnerAllTilesLevel21to30-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel31to40.json --experiment LodeRunnerAllTilesLevel31to40-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel41to50.json --experiment LodeRunnerAllTilesLevel41to50-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel51to60.json --experiment LodeRunnerAllTilesLevel51to60-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel61to70.json --experiment LodeRunnerAllTilesLevel61to70-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel71to80.json --experiment LodeRunnerAllTilesLevel71to80-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel81to90.json --experiment LodeRunnerAllTilesLevel81to90-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel91to100.json --experiment LodeRunnerAllTilesLevel91to100-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel101to110.json --experiment LodeRunnerAllTilesLevel101to110-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel111to120.json --experiment LodeRunnerAllTilesLevel111to120-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel121to130.json --experiment LodeRunnerAllTilesLevel121to130-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel131to140.json --experiment LodeRunnerAllTilesLevel131to140-5000_10_7 --tiles 7
python main.py --niter 5000 --nz 10 --json LodeRunnerAllTilesLevel141to150.json --experiment LodeRunnerAllTilesLevel141to150-5000_10_7 --tiles 7

Training sets are pushed to the repo and here are the commands to train on for all of the new training sets. 
@schrum2

**Comment by kdsteckel** [comment]
Next, I am going to start to calculate them for all 150 levels and put them into a chart. Right now I am testing it using just level 1 and all values seem correct. I think the percent of backtracking for the A* path may have to be calculated in the method with the simpleAStarDistance and the connectivity because it will need the mostRecentVisited as well. I also need to refactor LodeRunnerLevelTask to use these methods instead of calculating them as they were before. I think that the AStarDistance and connectivity will need to be calculated in the old way in that class though because we need mostRecentVisited, actionSequence, and the start state to get the AStarPath visualization for that class, or at the very least we would need to run the search in the CommonConstants.watch if statement to have access to those. 
@schrum2

**Comment by schrum2** [comment]
There is a data structure called Triple that can be used to return three values. Definitely refactor LodeRunnerLevelTask to call the same methods. How about instead of having a method directly return the three scores, you just have it return the A* path and the visited states in a Pair, and then new separate methods can compute the connectivity and the backtracking by taking these precomputed values as parameters.

**Comment by kdsteckel** [comment]
CSV file for analysis is now in the repo. @schrum2

**Comment by schrum2** [comment]
There are many obvious errors in the CSV file. The A* length is sometimes -1, but all levels should be beatable ... increase the a* budget enough to beat all levels.

Another issue that might not be an error, but is worth checking because it is suspicious: the percentSolid is 0 in several levels ... it looks like these levels consist entirely of diggable ground. This makes sense, but please confirm.

**Comment by kdsteckel** [comment]
In some cases the -1 for A* length may be the A* budget isn't high enough, but for some like level 4 it is an occurrence of the reproducible error we found so it may be a problem with the A* path model. Also, I'm pretty sure there are quite a few levels that have only diggable ground and no solid ground, but I will check this tomorrow.

**Comment by schrum2** [comment]
I thought we had previously verified that all of the levels from the original game were beatable, but if not then that makes solving #528 even more important. However, that issue should also be easier to troubleshoot if there is a Level from the original game failing ... easy to test the model after each tweak.

**Comment by schrum2** [comment]
This issue has been on hold since you've been working on #528 , but even though we don't have A* paths, we do have lots of other information. It makes me wonder if we can start clustering the levels in certain ways to make better training sets. I'm also thinking about the conditional GAN issue #538 that @cappsb is working on. Our current problem with Lode Runner is that we have too many levels of varying sorts, but maybe if we could come up with class labels for the levels, we could train a cGAN for them as well to get better results ... I'm just not sure with the class labels for Lode Runner levels should be yet. Look at the level images, and the spreadsheet, and maybe even browse people's opinions online, to see if we can come up with a way of categorizing Lode Runner levels into classes. Put some brainstorming ideas in this issue thread.

**Comment by kdsteckel** [comment]
These are the calculations that we found to analyze the original 150 levels: 
Percentage of empty tiles
Percentage of ladder tiles
Percentage of ground tiles (solid or diggable)
Percentage of solid tiles
Percentage of diggable tiles
Connectivity score
Length of A* path
Percent backtracking in A* path (this is the percentage of locations on the A* path that are visited more than once)

Some ideas for class labels for categorizing Lode Runner Levels: 

1. Two labels based on each percentage, one for over 80% and one for under 20% to try to get the extremes. There are a lot of levels that are either almost all one tile type or have almost none of that tile type. 
2. There were no levels in the analysis that had 100% connectivity, but we could use like the top 10 levels with the highest connectivity score for one label. 
3. In the analysis sheet, the longest valid A* path length is 318. We could have one label for an A* path length of over 300. But this label will most likely be changed once the sheet is filled. 
4. There are a lot of levels that have very distinct diagonal patterns with the ground tiles, so I think we should definitely group some of those together to see if we can generate those patterns. 
5. There are also a lot of levels that have words that are written out in the tiles, it might be cool to train a model on those levels to see the results. I doubt anything crazy would happen but it could be interesting. 
6. There are a lot of levels that require digging to be able to collect all the gold. We should have a model that uses all of those levels. 

Those are just a few ideas that I think we could use for class labels. How many did you want to use? @schrum2

**Comment by schrum2** [comment]
I see about 10 levels that have clear writing or letters/characters in them. I agree that it would be nice to have them in a training set ... go ahead and create that json and give me a training command.

I think that having a distinction between levels that do and don't require lots of digging could also be helpful, but I want to wait until we have working path calculations to measure that. In fact, once we have the TSP in #528 working, you are going to re-run this analysis, but add the following calculation: percentage of steps on the solution path that go through diggable ground. That should indicate an interesting category of level.

Some of these other ideas do represent training sets that might be worth creating, but lets hold off for now. If we actually apply the conditional GAN #538 then we will need to have a way of classifying every level into a distinct category, meaning no overlap between groups. That would be easy to do based on things like ladder or ground percentage, but I'm not sure that's the best distinction. Then again, maybe we just have to try lots of things and see what happens.

In fact, do the following as well:
Sort the levels in order of ladder percentage, and then divide them up into 10 groups (15 levels per group): 15 highest ladder percentages, next 15, next 15, ... , 15 lowest ladder percentages. Create those jsons and give me those training commands as well.

**Comment by schrum2** [comment]
Some reminders about this issue:

> I see about 10 levels that have clear writing or letters/characters in them. I agree that it would be nice to have them in a training set

Go ahead and train this model yourself, and commit it to the repo.

If #528 is really solved, then redo the analysis as described:

> you are going to re-run this analysis, but add the following calculation: percentage of steps on the solution path that go through diggable ground. That should indicate an interesting category of level.

Also, add the following calculations to the analysis:
- Total number of visited states by TSP. Remember that A* has to be run a log of times before TSP is used. The total cost is the SUM of the number of visited states across all of these calculations (revisiting states in different A* runs counts multiple times).
- Note that the above is different from the number of visited states across all of the A* runs. That should also be tracked, but is calculated by getting the union across the visited states from all runs, and logging the size of that final set.

Do this too:

> In fact, do the following as well:
Sort the levels in order of ladder percentage, and then divide them up into 10 groups (15 levels per group): 15 highest ladder percentages, next 15, next 15, ... , 15 lowest ladder percentages. Create those jsons

However, I think you can train the models now.

**Comment by schrum2** [comment]
In addition to the reminders above, recent work in #528 has revealed that some levels are unbeatable if you don't account for the enemy bringing gold to the player. Create a training set that has all of the levels that do not have this problem.

**Comment by kdsteckel** [comment]
The level anaylsis gets stalled out at level 75 and eventually returns an exception and stops. I am not sure why this is happening. Can you take a look when you get a chance? @schrum2

**Comment by kdsteckel** [comment]
![image](https://user-images.githubusercontent.com/50111166/86280804-22227100-bba2-11ea-8073-08bace928531.png)
This is the error I get from level 75. @schrum2

**Comment by kdsteckel** [comment]
If #528 is really solved, then redo the analysis as described:

you are going to re-run this analysis, but add the following calculation: percentage of steps on the solution path that go through diggable ground. That should indicate an interesting category of level.

Also, add the following calculations to the analysis:

Total number of visited states by TSP. Remember that A* has to be run a log of times before TSP is used. The total cost is the SUM of the number of visited states across all of these calculations (revisiting states in different A* runs counts multiple times).
Note that the above is different from the number of visited states across all of the A* runs. That should also be tracked, but is calculated by getting the union across the visited states from all runs, and logging the size of that final set.
Do this too:

In fact, do the following as well:
Sort the levels in order of ladder percentage, and then divide them up into 10 groups (15 levels per group): 15 highest ladder percentages, next 15, next 15, ... , 15 lowest ladder percentages. Create those jsons

However, I think you can train the models now.

**Comment by kdsteckel** [comment]
Percent Dug: 20% with 30 levels in each group

Have methods that create the different training sets generically.
